Add decoders config and example (#15981)
pingcap/docs: tidb-cloud/serverless-driver.md

}
```

> ### Custom column type decoders

Reviewer: I think we can put this example in the `Configure the serverless driver` section

In case you want change the returning column value format, you could provide a decoder config to `connect` method:

```ts
import { connect, ColumnType } from '@tidbcloud/serverless';

const conn = connect({
  url: 'mysql://[username]:[password]@[host]/[database]',
  decoders: {
    // serverless driver returns BIGINT field as text value by default, with this decoder you could change it to js builtin BigInt type.
    [ColumnType.BIGINT]: (rawValue: string) => BigInt(rawValue),
    
    // serverless driver return DATETIME field as text in 'yyyy-MM-dd HH:mm:ss' format, with this decoder you could transform the text value to js native Date object.
    [ColumnType.DATETIME]: (rawValue: string) => new Date(rawValue),
  }
})

> // You could also provide decoders config in execute method to override the connection config

Reviewer: ```suggestion
// You could also provide decoders config in SQL level to override the decoders in connection config
```

| `url`        | string   | N/A           | The URL for the database, in the `mysql://[username]:[password]@[host]/[database]` format, where `database` can be skipped if you intend to connect to the default database.                                                     |
| `fetch`      | function | global fetch  | Custom fetch function. For example, you can use the `undici` fetch in node.js.                                                                                                                                                   |
| `arrayMode`  | bool     | `false`       | Whether to return results as arrays instead of objects. To get better performance, set it to `true`.                                                                                                                             |
| `fullResult` | bool     | `false`       | Whether to return full result object instead of just rows. To get more detailed results, set it to `true`.                                                                                                                       |

> | `decoders`   | object   | `{}`          | The `decoders` object is a collection of key-value pairs, which enables you to customize the decoding process for different column types. In each pair, you can specify a column type as the key and specify a corresponding function as the value. This function takes the raw string value received from TiDB Cloud serverless driver as an argument and returns the decoded value. |

Reviewer 2: ```suggestion
| `decoders`   | object   | `{}`          | A collection of key-value pairs, which enables you to customize the decoding process for different column types. In each pair, you can specify a column type as the key and specify a corresponding function as the value. This function takes the raw string value received from TiDB Cloud serverless driver as an argument and returns the decoded value. |
```

At the SQL level, you can configure the following options:

| Option       | Type   | Default value | Description                                                                                                        |
|--------------|--------|---------------|--------------------------------------------------------------------------------------------------------------------|
| `arrayMode`  | bool   | `false`       | Whether to return results as arrays instead of objects. To get better performance, set it to `true`.               |
| `fullResult` | bool   | `false`       | Whether to return full result object instead of just rows. To get more detailed results, set it to `true`.         |

> | `decoders`   | object | `{}`          | The `decoders` object is a collection of key-value pairs, which enables you to customize the decoding process for different column types. In each pair, you can specify a column type as the key and specify a corresponding function as the value. This function takes the raw string value received from TiDB Cloud serverless driver as an argument and returns the decoded value. If you have configured `decoders` at both the connection and SQL levels, the key-value pairs with different keys configured at the connection level will be merged to the SQL level to take effect. If the same key (this is, column type) is specified at both levels, the value at the SQL level takes precedence. |

Reviewer 2: ```suggestion
| `decoders`   | object | `{}`          | A collection of key-value pairs, which enables you to customize the decoding process for different column types. In each pair, you can specify a column type as the key and specify a corresponding function as the value. This function takes the raw string value received from TiDB Cloud serverless driver as an argument and returns the decoded value. If you have configured `decoders` at both the connection and SQL levels, the key-value pairs with different keys configured at the connection level will be merged to the SQL level to take effect. If the same key (this is, column type) is specified at both levels, the value at the SQL level takes precedence. |
```

const tx = await conn.begin({isolation:"READ COMMITTED"})
```

### Customize column type decoders

> To customize the format of returned column values, you can configure the `decoder` option in the `connect` method as follows:

Reviewer 2: ```suggestion
To customize the format of returned column values, you can configure the `decoder` option in the `connect()` method as follows:
```

```ts
import { connect, ColumnType } from '@tidbcloud/serverless';

const conn = connect({
  url: 'mysql://[username]:[password]@[host]/[database]',
  decoders: {

>     // TiDB Cloud serverless driver returns the `BIGINT` type as text value by default. With this decoder, you can convert it to the JavaScript builtin `BigInt` type.

Reviewer 2: ```suggestion
    // By default, TiDB Cloud serverless driver returns the BIGINT type as text value. This decoder converts BIGINT to the JavaScript built-in BigInt type.
```

    [ColumnType.BIGINT]: (rawValue: string) => BigInt(rawValue),
    

>     // TiDB Cloud serverless driver returns the `DATETIME` type as the text value in the 'yyyy-MM-dd HH:mm:ss' format. With this decoder, you can convert the text value to the JavaScript native `Date` object.

Reviewer 2: ```suggestion
    // TiDB Cloud serverless driver returns the DATETIME type as the text value in the 'yyyy-MM-dd HH:mm:ss' format. This decoder converts the DATETIME text to the JavaScript native Date object.
```

    [ColumnType.DATETIME]: (rawValue: string) => new Date(rawValue),
  }
})

> // You can also configure the `decoder` option at the SQL level to override the decoders with the same keys at the connection level.

Reviewer 2: ```suggestion
// You can also configure the decoder option at the SQL level to override the decoders with the same keys at the connection level.
```

At the SQL level, you can configure the following options:

| Option       | Type   | Default value | Description                                                                                                        |
|--------------|--------|---------------|--------------------------------------------------------------------------------------------------------------------|
| `arrayMode`  | bool   | `false`       | Whether to return results as arrays instead of objects. To get better performance, set it to `true`.               |
| `fullResult` | bool   | `false`       | Whether to return full result object instead of just rows. To get more detailed results, set it to `true`.         |

> | `decoders`   | object | `{}`          | Same as connection level `decoders` option. (Will merge and override but not replace the connection level config.) |

Reviewer: ```suggestion
| `decoders`   | object | `{}`          | Custom decoders config for column types. The object key should be the the column type name. The value should be a function taking the raw string value received by serverless driver as argument, and returning the decoded value. (It will merge the decoder in the connection level configurations for the diffierent key and replace for the same key) |
```

const tx = await conn.begin({isolation:"READ COMMITTED"})
```

> ### Custom column type decoders

Reviewer 3: ```suggestion
### Customize column type decoders
```

> In case you want change the returning column value format, you could provide a decoder config to `connect` method:

Reviewer 3: ```suggestion
To customize the format of returned column values, you can configure the `decoder` option in the `connect` method as follows:
```

```ts
import { connect, ColumnType } from '@tidbcloud/serverless';

const conn = connect({
  url: 'mysql://[username]:[password]@[host]/[database]',
  decoders: {

>     // serverless driver returns BIGINT field as text value by default, with this decoder you could change it to js builtin BigInt type.

Reviewer 3: ```suggestion
    // TiDB Cloud serverless driver returns the `BIGINT` type as text value by default. With this decoder, you can convert it to the JavaScript builtin `BigInt` type.
```

    [ColumnType.BIGINT]: (rawValue: string) => BigInt(rawValue),
    

>     // serverless driver return DATETIME field as text in 'yyyy-MM-dd HH:mm:ss' format, with this decoder you could transform the text value to js native Date object.

Reviewer 3: ```suggestion
    // TiDB Cloud serverless driver returns the `DATETIME` type as the text value in the 'yyyy-MM-dd HH:mm:ss' format. With this decoder, you can convert the text value to the JavaScript native `Date` object.
```

    [ColumnType.DATETIME]: (rawValue: string) => new Date(rawValue),
  }
})

> // You could also provide decoders config in SQL level to override the decoders in connection config

Reviewer 3: ```suggestion
// You could also configure the `decoder` option at the SQL level to override the decoders with the same keys at the connection level.
```
